### PR TITLE
koordlet: refactor becpu resource store

### LINKED ISSUE #1450: [proposal]refactor node BE(BestEffort) CPU metircs cache to tsdb
**What is your proposal**:
see https://github.com/koordinator-sh/koordinator/issues/1241 for full description.

**Why is this needed**:
refactor node BE(BestEffort) CPU  metircs cache to tsdb for better performance and expansibility.

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by zshmmm**
### Ⅰ. Describe what this PR does

1. refactor node BE(BestEffort) CPU metircs cache to tsdb for better performance and expansibility.
2. see https://github.com/koordinator-sh/koordinator/issues/1241 for full description.

### Ⅱ. Does this pull request fix one issue?
fixes #1450

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/webhook/util/writer/atomic/atomic_writer.go [hunk 1]
```diff
@@ -85,15 +85,18 @@ const (
 // The Write algorithm is:
 //
 //  1. The payload is validated; if the payload is invalid, the function returns
-//  2. The current timestamped directory is detected by reading the data directory
+//     2. The current timestamped directory is detected by reading the data directory
```
Comment (saintube) [inline pkg/webhook/util/writer/atomic/atomic_writer.go:88]: Perhaps there is no need to change the descheduler, scheduler, and webhook packages.

File: pkg/descheduler/utils/sorter/helper.go [hunk 2]
```diff
@@ -24,10 +24,9 @@ import (
 
 // CompareFn compares p1 and p2 and returns:
 //
-//   -1 if p1 <  p2
-//    0 if p1 == p2
-//   +1 if p1 >  p2
-//
+//	-1 if p1 <  p2
```
Comment (eahydra) [inline pkg/descheduler/utils/sorter/helper.go:27]: Please rollback unrelated changes.

File: pkg/scheduler/plugins/elasticquota/core/decorator.go [hunk 3]
```diff
@@ -30,7 +30,8 @@ func RegisterDecorator(decorator Decorator) {
 
 // Decorator can decorate the Node, Pod and ElasticQuota as needed
 // e.g: If you want to append some additional resources into Node, Pod and ElasticQuota
-//  such as GPU Resources or modify Spec by GPU model, you can implement the Decorator.
```
Comment (eahydra) [inline pkg/scheduler/plugins/elasticquota/core/decorator.go:33]: and here also need to be rollbacked.

File: pkg/webhook/util/writer/atomic/atomic_writer.go (hunk 1 above)
Comment (eahydra) [reply pkg/webhook/util/writer/atomic/atomic_writer.go:88]: +1

**Review by eahydra — CHANGES_REQUESTED** [review]

File: pkg/koordlet/resmanager/cpu_evict.go [hunk 4]
```diff
@@ -97,55 +97,130 @@ func (c *CPUEvictor) cpuEvict() {
...
+	avgBECPUUsage, count := getBECPUMetric(metriccache.BEResouceAllocationUsage, querier, queryparam.Aggregate)
+	// BECPURequest
+	avgBECPURequest, _ := getBECPUMetric(metriccache.BEResouceAllocationRequest, querier, queryparam.Aggregate)
+	// BECPULimit
+	avgBECPULimit, _ := getBECPUMetric(metriccache.BEResouceAllocationRealLimit, querier, queryparam.Aggregate)
 
-	if !isBECPUUsageHighEnough(avgBECPUQueryResult.Metric, thresholdConfig.CPUEvictBEUsageThresholdPercent) {
-		klog.Warningf("cpuEvict by ResourceSatisfaction skipped,avg cpuUsage not Enough! metric: %+v", avgBECPUQueryResult.Metric)
-		return nil, 0
+	if !isAvgQueryResultValid(windowSeconds, c.resmanager.collectResUsedIntervalSeconds, count) {
```
Comment (zwzhang0107) [inline pkg/koordlet/resmanager/cpu_evict.go:115]: better to use min(count1, count2, count3)

File: pkg/koordlet/resmanager/cpu_evict.go [hunk 5]
```diff
@@ -97,55 +97,130 @@ func (c *CPUEvictor) cpuEvict() {
...
-	if !isAvgQueryResultValid(avgBECPUQueryResult, windowSeconds, c.resmanager.collectResUsedIntervalSeconds) {
-		return nil, 0
+	queryparam := generateQueryParamsAvg(windowSeconds)
+	querier, err := c.resmanager.metricCache.Querier(*queryparam.Start, *queryparam.End)
+	if err != nil {
+		klog.Warningf("get query failed, error %v", err)
+		return 0
 	}
+	// BECPUUsage
+	avgBECPUUsage, count := getBECPUMetric(metriccache.BEResouceAllocationUsage, querier, queryparam.Aggregate)
```
Comment (zwzhang0107) [inline pkg/koordlet/resmanager/cpu_evict.go:109]: milli value should be mentioned in name, e.g. avgBECPUMilliUsage
or use CPU Core as the unit

File: pkg/koordlet/resmanager/cpu_evict.go [hunk 6]
```diff
@@ -97,55 +97,130 @@ func (c *CPUEvictor) cpuEvict() {
...
+	if err != nil {
+		klog.Warningf("get query failed, error %v", err)
+		return 0
+	}
+	// BECPUUsage
+	currentBECPUUsage, _ := getBECPUMetric(metriccache.BEResouceAllocationUsage, querier, queryparam.Aggregate)
+	// BECPURequest
+	currentBECPURequest, _ := getBECPUMetric(metriccache.BEResouceAllocationRequest, querier, queryparam.Aggregate)
+	// BECPULimit
+	currentBECPULimit, _ := getBECPUMetric(metriccache.BEResouceAllocationRealLimit, querier, queryparam.Aggregate)
```
Comment (zwzhang0107) [inline pkg/koordlet/resmanager/cpu_evict.go:142]: xxxRealLimit

File: pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector.go [hunk 7]
```diff
@@ -151,7 +161,7 @@ func (b *beResourceCollector) getBECPURealMilliLimit() (int, error) {
 	return limit, nil
 }
 
-func (b *beResourceCollector) getBECPURequestSum() resource.Quantity {
+func (b *beResourceCollector) getBECPURequestMilliSum() int64 {
```
Comment (saintube) [inline pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector.go:164]: how about `getBECPURequestMilliCores`

File: pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector_test.go [hunk 8]
```diff
@@ -74,51 +74,39 @@ func Test_collectBECPUResourceMetric(t *testing.T) {
...
-	}
-
-	got := collector.metricDB.GetBECPUResourceMetric(params)
-	gotMetric := got.Metric
-
-	assert.Equal(t, int64(1500), gotMetric.CPURequest.MilliValue(), "checkRequest")
-	assert.Equal(t, int64(4), gotMetric.CPUUsed.Value(), "checkUsage")
-	assert.Equal(t, int64(8), gotMetric.CPURealLimit.Value(), "checkLimit")
+	assert.NotPanics(t, func() {
+		collector.collectBECPUResourceMetric()
```
Comment (saintube) [inline pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector_test.go:78]: could we keep the value checks?

File: pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector.go [hunk 9]
```diff
@@ -88,31 +87,42 @@ func (b *beResourceCollector) collectBECPUResourceMetric() {
...
+	beLimit, err01 := metriccache.NodeBEMetric.GenerateSample(
+		metriccache.MetricPropertiesFunc.NodeBE(string(metriccache.BEResourceCPU), string(metriccache.BEResouceAllocationRealLimit)), collectTime, float64(realMilliLimit))
+	beRequest, err02 := metriccache.NodeBEMetric.GenerateSample(
+		metriccache.MetricPropertiesFunc.NodeBE(string(metriccache.BEResourceCPU), string(metriccache.BEResouceAllocationRequest)), collectTime, float64(beCPUMilliRequest))
+
+	beUsage, err03 := metriccache.NodeBEMetric.GenerateSample(
+		metriccache.MetricPropertiesFunc.NodeBE(string(metriccache.BEResourceCPU), string(metriccache.BEResouceAllocationUsage)), collectTime, float64(beCPUUsageMilliCores))
+
+	if err01 != nil || err02 != nil || err03 != nil {
+		klog.Errorf("failed to collect node BECPU, err01: %v, err02: %v, err03: %v", err01, err02, err03)
```
Comment (saintube) [inline pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector.go:108]: Please revise the variable naming and logging for readability and self-defined.
`err01` is not a recommended naming, and the current log message is only understandable to the developers who have the codebase.
Same below.

File: pkg/koordlet/resmanager/cpu_evict.go [hunk 10]
```diff
@@ -306,3 +317,59 @@ func isSatisfactionConfigValid(thresholdConfig *slov1alpha1.ResourceThresholdStr
...
+		klog.Warningf("cpuEvict by ResourceSatisfaction skipped, queryResult %s error: %v", aggregateType, err)
+		return 0.0, 0
+	}
+	count := result.Count()
+
+	return value, int64(count)
+
+}
+
+func doQuery(querier metriccache.Querier, resource metriccache.MetricResource, properties map[metriccache.MetricProperty]string) (metriccache.AggregateResult, error) {
```
Comment (saintube) [inline pkg/koordlet/resmanager/cpu_evict.go:351]: This can be a common function for resmanager. Perhaps it should be moved into `resmanager/resmanager.go`.

File: pkg/koordlet/resmanager/cpu_evict.go [hunk 11]
```diff
@@ -306,3 +317,45 @@ func isSatisfactionConfigValid(thresholdConfig *slov1alpha1.ResourceThresholdStr
...
+		klog.Warningf("cpuEvict by ResourceSatisfaction skipped, queryResult %s error: %v", aggregateType, err)
+		return 0.0, 0
+	}
+	count := result.Count()
+
+	return value, int64(count)
+
+}
+
+func minCount(num ...int64) int64 {
```
Comment (zwzhang0107) [inline pkg/koordlet/resmanager/cpu_evict.go:351]: minInt64

**Review by eahydra — APPROVED** [review]
